synchronization/syncList.c

```diff
@@ -0,0 +1,200 @@
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <time.h>
+
+typedef struct _Node {
+    char value[100];
+    struct _Node* next;
+    pthread_mutex_t sync;
+} Node;
+
+typedef struct _Storage {
+    Node* first;
+} Storage;
+
+int increasing_length_count = 0;
+int decreasing_length_count = 0;
+int equal_length_count = 0;
+int swap_count = 0;
+
+void addNode(Storage* storage, const char* value) {
+    Node* new_node = (Node* )malloc(sizeof(Node));
+    strncpy(new_node->value, value, 100);
+    new_node->value[99] = '\0'; // Обеспечиваем, что строка будет нуль-терминированной
+    pthread_mutex_init(&new_node->sync, NULL);
+    new_node->next = storage->first;
+    storage->first = new_node;
+}
+
+void deleteNode(Storage* storage, Node* node) {
+    if (storage->first == NULL || node == NULL) return;
+
+    if (storage->first == node) {
+        storage->first = node->next;
+    } else {
+        Node* temp = storage->first;
+        while (temp->next != NULL && temp->next != node) {
+            temp = temp->next;
+        }
+        if (temp->next == node) {
+            temp->next = node->next;
+        }
+    }
+    pthread_mutex_destroy(&node->sync);
+    free(node);
+}
+
+// Функция потока для подсчета пар строк, идущих по возрастанию длины
+void* countIncreasingLengthPairs(void* arg) {
+    Storage* storage = (Storage*)arg;
+    while (1) {
+        pthread_mutex_lock(&storage->first->sync);
```

Comment:
Race preserved. When you get mutex to lock  (storage->first->sync and before actual lock your swapping thread might switch first element  and you'll lock mutex not for the first element.

Reply:
fixed